### PR TITLE
Named assembly labels.

### PR DESCRIPTION
**Pull request by chriseth**

### CODE REVIEW

File: libjulia/backends/evm/EVMAssembly.h [hunk 1]
```diff
@@ -52,6 +52,7 @@ class EVMAssembly: public AbstractAssembly
 	virtual void appendLabelReference(LabelID _labelId) override;
 	/// Generate a new unique label.
 	virtual LabelID newLabelId() override;
+	virtual LabelID namedLabel(std::string const&_name) override;
```
Comment (axic) [inline libjulia/backends/evm/EVMAssembly.h]: Should have documentation.

File: libevmasm/Assembly.cpp [hunk 2]
```diff
@@ -49,6 +49,8 @@ void Assembly::append(Assembly const& _a)
 	}
 	m_deposit = newDeposit;
 	m_usedTags += _a.m_usedTags;
+	// This does not append the named tags on purpose, they are only available by name inside
+	// the assembly.
```
Comment (axic) [inline libevmasm/Assembly.cpp]: Hm? I think it does. The tag for a named tag is just `newTag()` == `m_usedTags++`.

Comment (chriseth) [reply libevmasm/Assembly.cpp]: Ok, let me rephrase to "does not retain the names of named tags on purpose (the tags themselves are of course carried over)"

**Comment by axic** [comment]
@chriseth please address those two comments and then it should be ready to merge

**Comment by chriseth** [comment]
Updated.

**Review by axic — APPROVED** [review]

**Comment by axic** [comment]
I've restarted the tests about 3 times but they keep failing with timeouts. Maybe something is actually broken?

**Comment by axic** [comment]
Rebased & pushed, maybe that will fix CI.

**Comment by axic** [comment]
Seems to be stalling on this fuzzer test: test_c7249fb97a97085ff40b5b7690a6b62cbfe8fa744f4d532962582464ae0a17e1.sol

```
                pragma experimental ABIEncoderV2;
                contract c {
                        uint8[] x;
                        uint16[] y;
                        uint24[] z;
                        uint24[][] w;
                        function test1() returns (uint8[]) {
                                for (uint i = 0; i < 101; ++i)
                                        x.push(uint8(i));
                                return x;
                        }
                        function test2() returns (uint16[]) {
                                for (uint i = 0; i < 101; ++i)
                                        y.push(uint16(i));
                                return y;
                        }
                        function test3() returns (uint24[]) {
                                for (uint i = 0; i < 101; ++i)
                                        z.push(uint24(i));
                                return z;
                        }
                        function test4() returns (uint24[][]) {
                                w.length = 5;
                                for (uint i = 0; i < 5; ++i)
                                        for (uint j = 0; j < 101; ++j)
                                                w[i].push(uint24(j));
                                return w;
                        }
                }
```

**Comment by axic** [comment]
Weird, the fuzzer stalls on it, but compiling is a simple exception:
```
Internal compiler error during compilation:
/Users/alex/Projects/solidity/libsolidity/interface/CompilerStack.cpp(688): Throw in function void dev::solidity::CompilerStack::compileContract(const dev::solidity::ContractDefinition &, map<const dev::solidity::ContractDefinition *, const eth::Assembly *> &)
Dynamic exception type: boost::exception_detail::clone_impl<dev::solidity::InternalCompilerError>
std::exception::what: Assembly exception for bytecode
[dev::tag_comment*] = Assembly exception for bytecode
```

**Comment by axic** [comment]
It fails on this one:
```
+			assertThrow(m_tagPositionsInBytecode[size_t(i.data())] == size_t(-1), AssemblyException, "Duplicate tag position.");
```

**Comment by axic** [comment]
solfuzzer goes through standardcompiler (via jsoncompiler) and has optimisations turned on. In that case it stalls in deduplication:
```
* thread #1, queue = 'com.apple.main-thread', stop reason = signal SIGSTOP
  * frame #0: 0x000000010000682a solfuzzer`boost::multiprecision::backends::cpp_int_base<256u, 256u, (boost::multiprecision::cpp_integer_type)0, (boost::multiprecision::cpp_int_check_type)0, void, false>::cpp_int_base(boost::multiprecision::backends::cpp_int_base<256u, 256u, (boost::multiprecision::cpp_integer_type)0, (boost::multiprecision::cpp_int_check_type)0, void, false> const&) + 42
    frame #1: 0x00000001000067f8 solfuzzer`boost::multiprecision::backends::cpp_int_backend<256u, 256u, (boost::multiprecision::cpp_integer_type)0, (boost::multiprecision::cpp_int_check_type)0, void>::cpp_int_backend(boost::multiprecision::backends::cpp_int_backend<256u, 256u, (boost::multiprecision::cpp_integer_type)0, (boost::multiprecision::cpp_int_check_type)0, void> const&) + 40
    frame #2: 0x00000001000067bd solfuzzer`boost::multiprecision::backends::cpp_int_backend<256u, 256u, (boost::multiprecision::cpp_integer_type)0, (boost::multiprecision::cpp_int_check_type)0, void>::cpp_int_backend(boost::multiprecision::backends::cpp_int_backend<256u, 256u, (boost::multiprecision::cpp_integer_type)0, (boost::multiprecision::cpp_int_check_type)0, void> const&) + 29
    frame #3: 0x000000010000678d solfuzzer`boost::multiprecision::number<boost::multiprecision::backends::cpp_int_backend<256u, 256u, (boost::multiprecision::cpp_integer_type)0, (boost::multiprecision::cpp_int_check_type)0, void>, (boost::multiprecision::expression_template_option)0>::number(boost::multiprecision::number<boost::multiprecision::backends::cpp_int_backend<256u, 256u, (boost::multiprecision::cpp_integer_type)0, (boost::multiprecision::cpp_int_check_type)0, void>, (boost::multiprecision::expression_template_option)0> const&) + 29
    frame #4: 0x000000010000616d solfuzzer`boost::multiprecision::number<boost::multiprecision::backends::cpp_int_backend<256u, 256u, (boost::multiprecision::cpp_integer_type)0, (boost::multiprecision::cpp_int_check_type)0, void>, (boost::multiprecision::expression_template_option)0>::number(boost::multiprecision::number<boost::multiprecision::backends::cpp_int_backend<256u, 256u, (boost::multiprecision::cpp_integer_type)0, (boost::multiprecision::cpp_int_check_type)0, void>, (boost::multiprecision::expression_template_option)0>&&) + 29
    frame #5: 0x000000010001688b solfuzzer`std::__1::shared_ptr<boost::multiprecision::number<boost::multiprecision::backends::cpp_int_backend<256u, 256u, (boost::multiprecision::cpp_integer_type)0, (boost::multiprecision::cpp_int_check_type)0, void>, (boost::multiprecision::expression_template_option)0> > std::__1::shared_ptr<boost::multiprecision::number<boost::multiprecision::backends::cpp_int_backend<256u, 256u, (boost::multiprecision::cpp_integer_type)0, (boost::multiprecision::cpp_int_check_type)0, void>, (boost::multiprecision::expression_template_option)0> >::make_shared<boost::multiprecision::number<boost::multiprecision::backends::cpp_int_backend<256u, 256u, (boost::multiprecision::cpp_integer_type)0, (boost::multiprecision::cpp_int_check_type)0, void>, (boost::multiprecision::expression_template_option)0>&>(boost::multiprecision::number<boost::multiprecision::backends::cpp_int_backend<256u, 256u, (boost::multiprecision::cpp_integer_type)0, (boost::multiprecision::cpp_int_check_type)0, void>, (boost::multiprecision::expression_template_option)0>&&&) + 2043
    frame #6: 0x0000000100015002 solfuzzer`dev::eth::AssemblyItem::AssemblyItem(dev::eth::AssemblyItemType, boost::multiprecision::number<boost::multiprecision::backends::cpp_int_backend<256u, 256u, (boost::multiprecision::cpp_integer_type)0, (boost::multiprecision::cpp_int_check_type)0, void>, (boost::multiprecision::expression_template_option)0>, dev::SourceLocation const&) + 386
    frame #7: 0x0000000100014e73 solfuzzer`dev::eth::AssemblyItem::AssemblyItem(dev::eth::AssemblyItemType, boost::multiprecision::number<boost::multiprecision::backends::cpp_int_backend<256u, 256u, (boost::multiprecision::cpp_integer_type)0, (boost::multiprecision::cpp_int_check_type)0, void>, (boost::multiprecision::expression_template_option)0>, dev::SourceLocation const&) + 35
    frame #8: 0x000000010007c78b solfuzzer`dev::eth::AssemblyItem::pushTag() const + 699
    frame #9: 0x000000010007eebd solfuzzer`dev::eth::BlockDeduplicator::deduplicate()::$_0::operator()(unsigned long, unsigned long) const + 653
    frame #10: 0x000000010007ec23 solfuzzer`bool std::__1::__invoke_void_return_wrapper<bool>::__call<dev::eth::BlockDeduplicator::deduplicate()::$_0&, unsigned long, unsigned long>(dev::eth::BlockDeduplicator::deduplicate()::$_0&&&, unsigned long&&, unsigned long&&) + 115
    frame #11: 0x000000010007ea99 solfuzzer`std::__1::__function::__func<dev::eth::BlockDeduplicator::deduplicate()::$_0, std::__1::allocator<dev::eth::BlockDeduplicator::deduplicate()::$_0>, bool (unsigned long, unsigned long)>::operator()(unsigned long&&, unsigned long&&) + 73
    frame #12: 0x00000001000806aa solfuzzer`std::__1::function<bool (unsigned long, unsigned long)>::operator()(unsigned long, unsigned long) const + 186
    frame #13: 0x0000000100080f1e solfuzzer`std::__1::__tree_node_base<void*>*& std::__1::__tree<unsigned long, std::__1::function<bool (unsigned long, unsigned long)>, std::__1::allocator<unsigned long> >::__find_equal<unsigned long>(std::__1::__tree_end_node<std::__1::__tree_node_base<void*>*>*&, unsigned long const&) + 414
    frame #14: 0x0000000100080802 solfuzzer`std::__1::pair<std::__1::__tree_iterator<unsigned long, std::__1::__tree_node<unsigned long, void*>*, long>, bool> std::__1::__tree<unsigned long, std::__1::function<bool (unsigned long, unsigned long)>, std::__1::allocator<unsigned long> >::__emplace_unique_key_args<unsigned long, unsigned long const&>(unsigned long const&, unsigned long const&&&) + 82
    frame #15: 0x000000010007be0d solfuzzer`dev::eth::BlockDeduplicator::deduplicate() + 2541
    frame #16: 0x000000010003a849 solfuzzer`dev::eth::Assembly::optimiseInternal(dev::eth::Assembly::OptimiserSettings const&, std::__1::set<unsigned long, std::__1::less<unsigned long>, std::__1::allocator<unsigned long> > const&) + 953
    frame #17: 0x000000010003a5d4 solfuzzer`dev::eth::Assembly::optimiseInternal(dev::eth::Assembly::OptimiserSettings const&, std::__1::set<unsigned long, std::__1::less<unsigned long>, std::__1::allocator<unsigned long> > const&) + 324
    frame #18: 0x000000010003a424 solfuzzer`dev::eth::Assembly::optimise(dev::eth::Assembly::OptimiserSettings const&) + 100
    frame #19: 0x000000010003a36f solfuzzer`dev::eth::Assembly::optimise(bool, bool, unsigned long) + 127
    frame #20: 0x000000010043fe79 solfuzzer`dev::solidity::CompilerContext::optimise(bool, unsigned int) + 73
    frame #21: 0x000000010043fd1c solfuzzer`dev::solidity::Compiler::compileContract(dev::solidity::ContractDefinition const&, std::__1::map<dev::solidity::ContractDefinition const*, dev::eth::Assembly const*, std::__1::less<dev::solidity::ContractDefinition const*>, std::__1::allocator<std::__1::pair<dev::solidity::ContractDefinition const* const, dev::eth::Assembly const*> > > const&, std::__1::vector<unsigned char, std::__1::allocator<unsigned char> > const&) + 332
    frame #22: 0x000000010059d5bf solfuzzer`dev::solidity::CompilerStack::compileContract(dev::solidity::ContractDefinition const&, std::__1::map<dev::solidity::ContractDefinition const*, dev::eth::Assembly const*, std::__1::less<dev::solidity::ContractDefinition const*>, std::__1::allocator<std::__1::pair<dev::solidity::ContractDefinition const* const, dev::eth::Assembly const*> > >&) + 5871
    frame #23: 0x000000010059bdc3 solfuzzer`dev::solidity::CompilerStack::compile() + 1235
    frame #24: 0x000000010062148b solfuzzer`dev::solidity::StandardCompiler::compileInternal(Json::Value const&) + 18843
    frame #25: 0x0000000100628ff4 solfuzzer`dev::solidity::StandardCompiler::compile(Json::Value const&) + 68
    frame #26: 0x00000001000225fe solfuzzer`(anonymous namespace)::compile(std::__1::map<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, std::__1::less<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > >, std::__1::allocator<std::__1::pair<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > > > > const&, bool, void (*)(char const*, char**, char**)) + 2478
    frame #27: 0x000000010001fbbd solfuzzer`(anonymous namespace)::compileSingle(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, bool) + 349
    frame #28: 0x000000010001f58c solfuzzer`compileJSON + 236
    frame #29: 0x0000000100004386 solfuzzer`(anonymous namespace)::testCompiler() + 502
    frame #30: 0x00000001000026f7 solfuzzer`main + 2327
    frame #31: 0x00007fff9227a235 libdyld.dylib`start + 1
```

**Comment by chriseth** [comment]
Oh wow, this sounds serious!

**Comment by chriseth** [comment]
Thanks for digging into this, I will try to fix it.

**Comment by chriseth** [comment]
Ah ok, stupid mistake. The fix is that this commit cannot be used on its own, only in combination with #2823

**Comment by axic** [comment]
What is the reason?

**Comment by chriseth** [comment]
There are functions with the same name until all of them are combined into the same assembly block.
